src/server/pkg/worker/api_server.go

```diff
@@ -385,8 +395,27 @@ func (a *APIServer) Process(ctx context.Context, req *ProcessRequest) (resp *Pro
 	// We cannot run more than one user process at once; otherwise they'd be
 	// writing to the same output directory. Acquire lock to make sure only one
 	// user process runs at a time.
-	a.Lock()
-	defer a.Unlock()
+	a.processMu.Lock()
+	defer a.processMu.Unlock()
+	// set the status for the datum
+	ctx, cancel := context.WithCancel(ctx)
+	func() {
+		a.statusMu.Lock()
+		defer a.statusMu.Unlock()
+		a.jobID = req.JobID
+		a.data = req.Data
+		a.started = time.Now()
+		a.cancel = cancel
+	}()
+	// unset the status when this function exits
+	defer func() {
+		a.statusMu.Lock()
+		defer a.statusMu.Unlock()
+		a.jobID = ""
+		a.data = nil
+		a.started = time.Time{}
+		a.cancel = nil
```

Comment:
So after trying to do this I'm not quite sure it makes things nicer, there's a couple of things that make it annoying.

- `cancel` isn't part of the protobuf struct, since it's a function, so we still need that one
- `data` isn't only used in status, it's also used to download inputs, furthermore the protobuf uses `Datum`s not `FileInfo`s so this actually gives us `2` fields called `data` that need to be maintained.
- we can't return the `WorkerStatus` object directly from the `Status` method because it contains pointers which might get modified before serialization happens. So we need to do a deep-copy before returning.

All in all it's feels like it's not quite worth it to me.